File: web/src/api/modules/catalog.ts
import apiService from '../api';
import type { ApiResponse, PaginatedResponse } from '../api';

export type CatalogTemplateStatus = 'draft' | 'pending_review' | 'published' | 'rejected';
export type CatalogTemplateVisibility = 'private' | 'public';
export type CatalogDeployTarget = 'k8s' | 'compose';

export interface CatalogVariableSchema {
  name: string;
  type: 'string' | 'number' | 'password' | 'boolean' | 'select' | 'textarea';
  default?: string | number | boolean;
  required: boolean;
  description?: string;
  options?: string[];
}

export interface CatalogCategory {
  id: number;
  name: string;
  display_name: string;
  icon?: string;
  description?: string;
  sort_order: number;
  is_system: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface CatalogTemplate {
  id: number;
  name: string;
  display_name: string;
  description?: string;
  icon?: string;
  category_id: number;
  version: string;
  owner_id: number;
  visibility: CatalogTemplateVisibility;
  status: CatalogTemplateStatus;
  k8s_template?: string;
  compose_template?: string;
  variables_schema: CatalogVariableSchema[];
  readme?: string;
  tags: string[];
  deploy_count: number;
  review_note?: string;
  created_at?: string;
  updated_at?: string;
}

export interface CategoryCreatePayload {
  name: string;
  display_name: string;
  icon?: string;
  description?: string;
  sort_order?: number;
}

export interface TemplateCreatePayload {
  name: string;
  display_name: string;
  description?: string;
  icon?: string;
  category_id: number;
  version?: string;
  visibility?: CatalogTemplateVisibility;
  k8s_template?: string;
  compose_template?: string;
  variables_schema: CatalogVariableSchema[];
  readme?: string;
  tags?: string[];
}

export interface CatalogPreviewPayload {
  template_id: number;
  target: CatalogDeployTarget;
  variables: Record<string, string | number | boolean>;
}

export interface CatalogPreviewResponse {
  rendered_yaml: string;
  unresolved_vars: string[];
}

export interface CatalogDeployPayload {
  template_id: number;
  target: CatalogDeployTarget;
  project_id: number;
  team_id?: number;
  service_name: string;
  namespace?: string;
  cluster_id?: number;
  environment?: string;
  variables: Record<string, string | number | boolean>;
  deploy_now?: boolean;
}

export interface CatalogDeployResponse {
  service_id: number;
  template_id: number;
  deploy_count: number;
}

export interface CatalogTemplateListParams {
  category_id?: number;
  status?: CatalogTemplateStatus;
  visibility?: CatalogTemplateVisibility;
  q?: string;
  mine?: boolean;
}

export const catalogApi = {
  async listCategories(): Promise<ApiResponse<PaginatedResponse<CatalogCategory>>> {
    const response = await apiService.get<any>('/catalog/categories');
    return {
      ...response,
      data: {
        list: Array.isArray(response.data?.list) ? response.data.list : [],
        total: Number(response.data?.total || 0),
      },
    };
  },

  async createCategory(payload: CategoryCreatePayload): Promise<ApiResponse<CatalogCategory>> {
    return apiService.post('/catalog/categories', payload);
  },

  async updateCategory(id: number, payload: Partial<CategoryCreatePayload>): Promise<ApiResponse<CatalogCategory>> {
    return apiService.put(`/catalog/categories/${id}`, payload);
  },

  async deleteCategory(id: number): Promise<ApiResponse<void>> {
    return apiService.delete(`/catalog/categories/${id}`);
  },

  async listTemplates(params?: CatalogTemplateListParams): Promise<ApiResponse<PaginatedResponse<CatalogTemplate>>> {
    const response = await apiService.get<any>('/catalog/templates', { params });
    return {
      ...response,
      data: {
        list: Array.isArray(response.data?.list) ? response.data.list : [],
        total: Number(response.data?.total || 0),
      },
    };
  },

  async getTemplate(id: number): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.get(`/catalog/templates/${id}`);
  },

  async createTemplate(payload: TemplateCreatePayload): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.post('/catalog/templates', payload);
  },

  async updateTemplate(id: number, payload: Partial<TemplateCreatePayload>): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.put(`/catalog/templates/${id}`, payload);
  },

  async deleteTemplate(id: number): Promise<ApiResponse<void>> {
    return apiService.delete(`/catalog/templates/${id}`);
  },

  async submitTemplate(id: number): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.post(`/catalog/templates/${id}/submit`);
  },

  async publishTemplate(id: number): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.post(`/catalog/templates/${id}/publish`);
  },

  async rejectTemplate(id: number, reason: string): Promise<ApiResponse<CatalogTemplate>> {
    return apiService.post(`/catalog/templates/${id}/reject`, { reason });
  },

  async preview(payload: CatalogPreviewPayload): Promise<ApiResponse<CatalogPreviewResponse>> {
    return apiService.post('/catalog/preview', payload);
  },

  async deploy(payload: CatalogDeployPayload): Promise<ApiResponse<CatalogDeployResponse>> {
    return apiService.post('/catalog/deploy', payload);
  },
};

export default catalogApi;

File: api/catalog/v1/catalog.go
package v1

import "time"

type CatalogVariableSchema struct {
	Name        string   `json:"name"`
	Type        string   `json:"type"`
	Default     any      `json:"default,omitempty"`
	Required    bool     `json:"required"`
	Description string   `json:"description,omitempty"`
	Options     []string `json:"options,omitempty"`
}

type CategoryCreateRequest struct {
	Name        string `json:"name"`
	DisplayName string `json:"display_name"`
	Icon        string `json:"icon"`
	Description string `json:"description"`
	SortOrder   int    `json:"sort_order"`
}

type CategoryUpdateRequest struct {
	DisplayName *string `json:"display_name,omitempty"`
	Icon        *string `json:"icon,omitempty"`
	Description *string `json:"description,omitempty"`
	SortOrder   *int    `json:"sort_order,omitempty"`
}

type CategoryResponse struct {
	ID          uint      `json:"id"`
	Name        string    `json:"name"`
	DisplayName string    `json:"display_name"`
	Icon        string    `json:"icon"`
	Description string    `json:"description"`
	SortOrder   int       `json:"sort_order"`
	IsSystem    bool      `json:"is_system"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type TemplateCreateRequest struct {
	Name            string                  `json:"name"`
	DisplayName     string                  `json:"display_name"`
	Description     string                  `json:"description"`
	Icon            string                  `json:"icon"`
	CategoryID      uint                    `json:"category_id"`
	Version         string                  `json:"version"`
	Visibility      string                  `json:"visibility"`
	K8sTemplate     string                  `json:"k8s_template"`
	ComposeTemplate string                  `json:"compose_template"`
	VariablesSchema []CatalogVariableSchema `json:"variables_schema"`
	Readme          string                  `json:"readme"`
	Tags            []string                `json:"tags"`
}

type TemplateUpdateRequest struct {
	DisplayName     *string                  `json:"display_name,omitempty"`
	Description     *string                  `json:"description,omitempty"`
	Icon            *string                  `json:"icon,omitempty"`
	CategoryID      *uint                    `json:"category_id,omitempty"`
	Version         *string                  `json:"version,omitempty"`
	Visibility      *string                  `json:"visibility,omitempty"`
	K8sTemplate     *string                  `json:"k8s_template,omitempty"`
	ComposeTemplate *string                  `json:"compose_template,omitempty"`
	VariablesSchema *[]CatalogVariableSchema `json:"variables_schema,omitempty"`
	Readme          *string                  `json:"readme,omitempty"`
	Tags            *[]string                `json:"tags,omitempty"`
}

type TemplateResponse struct {
	ID              uint                    `json:"id"`
	Name            string                  `json:"name"`
	DisplayName     string                  `json:"display_name"`
	Description     string                  `json:"description"`
	Icon            string                  `json:"icon"`
	CategoryID      uint                    `json:"category_id"`
	Version         string                  `json:"version"`
	OwnerID         uint64                  `json:"owner_id"`
	Visibility      string                  `json:"visibility"`
	Status          string                  `json:"status"`
	K8sTemplate     string                  `json:"k8s_template"`
	ComposeTemplate string                  `json:"compose_template"`
	VariablesSchema []CatalogVariableSchema `json:"variables_schema"`
	Readme          string                  `json:"readme"`
	Tags            []string                `json:"tags"`
	DeployCount     int                     `json:"deploy_count"`
	ReviewNote      string                  `json:"review_note"`
	CreatedAt       time.Time               `json:"created_at"`
	UpdatedAt       time.Time               `json:"updated_at"`
}

type TemplateListResponse struct {
	List  []TemplateResponse `json:"list"`
	Total int64              `json:"total"`
}

type ReviewActionRequest struct {
	Reason string `json:"reason"`
}

type PreviewRequest struct {
	TemplateID uint           `json:"template_id"`
	Target     string         `json:"target"`
	Variables  map[string]any `json:"variables"`
}

type PreviewResponse struct {
	RenderedYAML   string   `json:"rendered_yaml"`
	UnresolvedVars []string `json:"unresolved_vars"`
}

type DeployRequest struct {
	TemplateID  uint           `json:"template_id"`
	Target      string         `json:"target"`
	ProjectID   uint           `json:"project_id"`
	TeamID      uint           `json:"team_id"`
	ServiceName string         `json:"service_name"`
	Namespace   string         `json:"namespace"`
	ClusterID   uint           `json:"cluster_id"`
	Environment string         `json:"environment"`
	Variables   map[string]any `json:"variables"`
	DeployNow   bool           `json:"deploy_now"`
}

type DeployResponse struct {
	ServiceID   uint `json:"service_id"`
	TemplateID  uint `json:"template_id"`
	DeployCount int  `json:"deploy_count"`
}
